[TAN-5839] Add project/folder exclusion to ProjectsWidget

AchrafGoVocal opened the pull request:
# Changelog
## Added
- Added projects/folder exclusion to "Projects Widget" and to "Projects Timeline Widget"

AchrafGoVocal committed:
[TAN-5839] Add project/folder exclusion to ProjectsWidget

Commit:
Translations updated by CI (extract-intl)

AchrafGoVocal committed:
[TAN-5839] Remove unused exports from ReportAdminPublicationSelector

AchrafGoVocal committed:
Merge branch 'TAN-5839-exclude-projects-from-widget' of github.com:CitizenLabDotCo/citizenlab into TAN-5839-exclude-projects-from-widget

AchrafGoVocal committed:
[TAN-5839] Improve ProjectsWidget settings

AchrafGoVocal committed:
[TAN-5839] Add specs for excluded_admin_publication_ids parameter

AchrafGoVocal committed:
[TAN-5839] Add excluded_admin_publication_ids filtering to ProjectsFinderAdminService

AchrafGoVocal committed:
[TAN-5839] Add project/folder exclusion to ProjectsTimelineWidget

Commit:
Translations updated by CI (extract-intl)

sebastienhoorens reviewed (approved):
All good on the backend side. I think it's best if a frontend developer reviews the frontend part because I'm don't follow everything too well there.

Is there only one field that filters both projects and folders? I think the main thing I would question is why this decision and not filtering projects and folders separately. But this is fine I think.

Also, could there be any situation where admins are confused because they added a folder to be filtered out, but the projects inside that folder themselves are not being filtered out (if I understand correctly)?

One more thing: When deleting a project/folder, it doesn't seem like the IDs will get cleared from where they are stored (I don't know where this is, perhaps some JSON column?). In that case, we might want to make sure that we don't get bugs with the report crashing because a project/folder referenced in a filter no longer exists. If these IDs are indeed persisted, perhaps it could be useful to add a test for this case?

AchrafGoVocal commented:
> All good on the backend side. I think it's best if a frontend developer reviews the frontend part because I'm don't follow everything too well there.
> 
> Is there only one field that filters both projects and folders? I think the main thing I would question is why this decision and not filtering projects and folders separately. But this is fine I think.
> 
> Also, could there be any situation where admins are confused because they added a folder to be filtered out, but the projects inside that folder themselves are not being filtered out (if I understand correctly)?
> 
> One more thing: When deleting a project/folder, it doesn't seem like the IDs will get cleared from where they are stored (I don't know where this is, perhaps some JSON column?). In that case, we might want to make sure that we don't get bugs with the report crashing because a project/folder referenced in a filter no longer exists. If these IDs are indeed persisted, perhaps it could be useful to add a test for this case?

**Re: Filtering projects and folders together:** I followed the UI design specified in the ticket, which combines projects and folders in a single filter. I agree that separating them could be clearer - I'll check with the product team. 

**Re: Folder exclusion and child projects:** The current implementation intentionally handles folder cascading on the frontend to give users more granular control. When a user excludes a folder:
- Frontend automatically selects all child project IDs within that folder
- These IDs are sent to the backend
- User can then manually un-exclude specific projects within that folder if needed

**Trade-off:**
- Current approach: More frontend complexity, but users have fine-grained control
- Backend approach: Simpler implementation, but less flexible (all-or-nothing exclusion)

**Re: Deleted projects/folders causing crashes:** You're right about this issue. The excluded IDs are persisted in the report widget's JSON configuration. I'll make sure to write a test for that case and also maybe look into finding a way of automatically cleaning up references to deleted `admin_publication`

AchrafGoVocal committed:
[TAN-5839] Add separate project and folder exclusion filters to ReportBuilder widgets

AchrafGoVocal committed:
[TAN-5839] Remove remnants of old exclusion logic

Commit:
Translations updated by CI (extract-intl)

AchrafGoVocal committed:
[TAN-5839] Move exclusion messages to shared messages

AchrafGoVocal committed:
[TAN-5839] Add specs for the new exclusion logic

AchrafGoVocal committed:
Merge branch 'TAN-5839-exclude-projects-from-widget' of github.com:CitizenLabDotCo/citizenlab into TAN-5839-exclude-projects-from-widget

Commit:
Translations updated by CI (extract-intl)

AchrafGoVocal committed:
[TAN-5839] Fix Rubocop offense

AchrafGoVocal committed:
Merge branch 'TAN-5839-exclude-projects-from-widget' of github.com:CitizenLabDotCo/citizenlab into TAN-5839-exclude-projects-from-widget

amanda-anderson reviewed (approved):
LGTM! :rocket:

AchrafGoVocal committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into TAN-5839-exclude-projects-from-widget

AchrafGoVocal committed:
[TAN-5839] Fix rubocop offense

AchrafGoVocal committed:
[TAN-5839] Add edge case specs for projects and project_timelines exclusion filters

sebastienhoorens reviewed (commented):
Do you think it could be possible to let ReportBuilder::Queries::Projects call ProjectsFinderAdminService? That would avoid the code duplication there seems to be at the moment.

AchrafGoVocal committed:
[TAN-5839] Refactor Projects query to use ProjectsFinderAdminService for all filtering

AchrafGoVocal commented:
> Do you think it could be possible to let ReportBuilder::Queries::Projects call ProjectsFinderAdminService? That would avoid the code duplication there seems to be at the moment.

Yes 👍 I think that makes sense. I may have gone a bit overboard with the refactor though 😬. I also reorganized some of the date range filtering tests to group them better and added a few edge cases.

I changed the behavior so that if there’s no overlap with the date range, it now returns an empty set. Previously it would return some projects, but it wasn’t clear which ones or why (especially with weird ranges like 1980–1981) so this felt less confusing.

AchrafGoVocal committed:
[TAN-5839] Clean up params passing

AchrafGoVocal committed:
[TAN-5839] Clean up useless tests

sebastienhoorens reviewed (approved):
All good 👍 Before merging, could you merge master into your branch and fix the offenses? match_array is now replaced by contain_exactly.

AchrafGoVocal committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into TAN-5839-exclude-projects-from-widget

AchrafGoVocal committed:
[TAN-5839] Fix match_array offenses
